src/go/k8s/internal/util/client.go

```diff
@@ -0,0 +1,353 @@
+package util
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-logr/logr"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/cluster.redpanda.com/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kgo"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceLabel                = "kubernetes.io/service-name"
+	helmInternalCertificateName = "default"
+	defaultCertSecretSuffix     = "-default-cert"
+
+	adminPort = "kafka"
+)
+
+type remapper interface {
+	MapHost(broker string) string
+	ConfigureTLS(brokers []string, config *tls.Config) *tls.Config
+}
+
+type noopRemapper struct{}
+
+func (n *noopRemapper) MapHost(broker string) string {
+	return broker
+}
+
+func (n *noopRemapper) ConfigureTLS(brokers []string, config *tls.Config) *tls.Config { return config }
+
+func defaultCertificate(name string) string {
+	return name + defaultCertSecretSuffix
+}
+
+type ClientFactory struct {
+	client.Client
+	dialer *net.Dialer
+
+	// this can be overridden in tests to use for port-forwarding into kubernetes
+	// installations from the host
+	remapper remapper
+}
+
+func NewClientFactory(client client.Client) *ClientFactory {
+	return &ClientFactory{
+		Client:   client,
+		remapper: &noopRemapper{},
+		dialer:   &net.Dialer{Timeout: 10 * time.Second},
+	}
+}
+
+// Admin returns a client able to communicate with the cluster defined by the given KafkaAPISpec.
+func (c *ClientFactory) Admin(ctx context.Context, namespace string, spec *redpandav1alpha1.KafkaAPISpec) (*kadm.Client, error) {
+	client, err := c.getClient(ctx, namespace, spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return kadm.NewClient(client), nil
+}
+
+// ClusterAdmin returns a client able to communicate with the given Redpanda cluster.
+func (c *ClientFactory) ClusterAdmin(ctx context.Context, cluster *redpandav1alpha2.Redpanda) (*kadm.Client, error) {
+	client, err := c.getClusterClient(ctx, adminPort, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return kadm.NewClient(client), nil
+}
+
+type dialer func(ctx context.Context, network string, host string) (net.Conn, error)
+
+func (c *ClientFactory) dialAndRemap(dialer dialer) dialer {
+	return func(ctx context.Context, network string, host string) (net.Conn, error) {
+		mapped := c.remapper.MapHost(host)
+		return dialer(ctx, network, mapped)
```

Comment:
One of my favorite abstractions I've ever written/used was a wrapper over the SPDY dialer that presented itself as a `net.Dialer`. It allowed dialing directly into Pods by their name without having to deal with extra ports or extra goroutines. Seems like we'd heavily benefit from implementing something similar as we can likely avoid this type of indirection.